File: labs/lab01_model_to_rf_tone/python/iq_tone_basic.py
#!/usr/bin/env python3
from pathlib import Path
import numpy as np
FS_HZ = 1.0e6
N = 4096
F0_HZ = 100.0e3
A = 0.9
PHI = np.pi/6
idx = np.arange(N, dtype=np.float64)
x = A * np.exp(1j * (2*np.pi*F0_HZ*idx/FS_HZ + PHI))
out = Path('results'); out.mkdir(exist_ok=True)
np.savetxt(out/'iq_tone_basic_iq.txt', np.column_stack((np.real(x), np.imag(x))), fmt='%.12f')
X = np.fft.fftshift(np.fft.fft(x))
f = np.fft.fftshift(np.fft.fftfreq(N, d=1.0/FS_HZ))
np.savetxt(out/'iq_tone_basic_fft.txt', np.column_stack((f, np.abs(X))), fmt='%.12f')
print('Generated IQ tone and saved reference files to results/.')

File: labs/lab01_model_to_rf_tone/cpp/iq_tone_basic.cpp
#include <cmath>
#include <complex>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>
namespace fs = std::filesystem;
int main() {
    const double Fs = 1.0e6, f0 = 100.0e3, A = 0.9, phi = M_PI/6.0;
    const std::size_t N = 4096;
    std::vector<std::complex<double>> x(N);
    for (std::size_t i = 0; i < N; ++i) {
        double p = 2.0 * M_PI * f0 * static_cast<double>(i) / Fs + phi;
        x[i] = A * std::exp(std::complex<double>(0.0, p));
    }
    fs::create_directories("results");
    std::ofstream os("results/iq_tone_basic_iq.txt");
    for (const auto& s : x) os << s.real() << ' ' << s.imag() << '
';
    std::cout << "Generated IQ tone and saved results/iq_tone_basic_iq.txt
";
}
